rep-2017.rst

```diff
@@ -0,0 +1,359 @@
+REP: 2017
+Title: Thread attributes configuration support in rcl
+Author: Masaaki Ueno, Koichi Oyama, Shoji Morita <s-morita@esol.co.jp>
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 11-Jul-2023
+Post-History: 18-Aug-2023
+
+
+Abstract
+========
+
+This REP proposed a feature of the ros client library (rcl) to accept the thread attributes, such as priorities, scheduling, and core affinity, which are to be passed to the language bindings like rclcpp.
+
+
+Motivation
+==========
+
+The configuration and setting of thread attributes for ROS 2 executor threads are essential for controlling computational resources and meeting real-time requirements, which are typically mandatory for robot systems.
+
+These attributes are particularly critical for embedded robot systems that lack redundant resources, as they aid in optimizing CPU resource utilization.
+
+However, currently, there is a lack of infrastructure to configure and set thread attributes for the threads used by the ROS 2 executors.  As a result, specific code tailored to each target environment must be implemented.  The implementation not only requires significant effort but poses some obstacles to continuous integration applied to the systems with multiple execution environments.
+
+Having an infrastructure that configures thread attributes for ROS 2 executor through command line parameters, environment variables, and/or files eliminates the need for additional implementation in each environment and simplifies the deployment process across multiple environments without requiring code modifications.
+
+The purpose of this REP is to propose a feature to the ros client library that allows it to accept thread attributes, store them internally, and provide interfaces to pass them to upper language bindings (e.g., rclcpp, rclpy, or future ones).  By incorporating this feature and enabling language bindings to read the thread attributes, ROS 2 can effectively adopt them for real-time systems or environments that require strict CPU resource optimization, eliminating the need for code modifications in each specific environment.
+
+Specification
+=============
+
+This chapter provides the interface specifications depicted as [Parameter configuration] and [Language binding] below.
+
+::
+
+                              +-----------+ +------------+
```

Comment:
is your proposal applicable for the `rclpy` as well?

Reply:
@razr 
> is your proposal applicable for the rclpy as well?

In a strict sense, it can be applied to rclpy. However, some additional code will be required to make it compatible with the newly-added rcl interface in rclpy. Additionally, we may need to implement an additional Python thread control class, similar to what our team has done with rcpputils.